연락처관리 UI [STEP 3, Bonus] Avery, Logan

devxsby opened the pull request:
> 안녕하세요 (@protocorn93) 콘 👨🏻‍🌾 
Avery와 Logan 이에요. 지난 한 주간 잘 지내셨나요 ~~? 😎
저번 Step2 리뷰도 너무 잘 봤어요! 감사해요! 🫶🏻 이번 Step3와 보너스는 요구사항 이외에 추가적으로 몇 개 더 구현해보았어요.
부족하지만 사용자를 생각하며 이런 기능이 있다면 더욱 편리하지 않을까?라는 것에 초점을 맞추고 진행하다보니, 미처 발견하지 못한 점도 있을 수 있고 부족한 부분도 있어요. 하지만 콘이 좋은 피드백 남겨주시면 반영해보도록 할게요!  🚀
늘 따뜻한 리뷰 달아주셔서 너무 감사하고 이번 스탭 리뷰도 잘 부탁드려요. 감사합니다 😊 🤍


## 📸  시뮬레이션

<img src="https://user-images.githubusercontent.com/80062632/218022661-0205032a-ced1-4f97-ac12-8223feffc435.gif" width="300px">

## 1️⃣ STEP 3 구현 사항

- [x]  연락처 목록 화면의 우상단 버튼을 통해 연락처 추가 화면에 진입할 수 있어요.
- [x]  오토 레이아웃과 스택뷰를 활용하여 화면을 레이아웃해요.
- [x] 각 필드에 맞는 키보드 종류를 지정해요.
    - 연락처 정보는 숫자 키보드를 사용하고 가운데 `-` 표시는 따로 사용자에게 입력받지 않고 처리해요.
- [x] 취소 버튼을 선택하면 정말 취소할 것인지 물어요.
- [x] 저장 버튼을 선택하면 입력한 정보가 올바르게 입력되었는지 확인해요.
  - 이름은 사용자가 중간에 띄어쓰기를 하더라도 띄어쓰기를 제거해요.
  - 나이는 숫자로만 입력해야하며, 세 자리수 이하여야 해요.
  - 연락처는 중간에 `-`로 구분하며, `-`는 두 개 존재해야 해요.
    - `-` 을 제외하고 숫자는 9자리 이상이어야 해요.

모든 정보가 올바로 입력되었다면 저장 버튼을 선택했을 때 이전 화면으로 돌아가고, 추가한 연락처가 연락처 리스트에 추가돼요.

## 2️⃣ STEP Bonus 구현 사항

- [x] 테이블뷰의 셀을 스와이프하여 삭제하는 메뉴를 보이고, 실제로 삭제할 수 있도록 구현해요.
- [x]  Cell을 사용자정의하여 원하는 레이아웃으로 표현하도록 해요.
- [x] 테이블 뷰 검색 기능을 구현해요.
- [x] Dynamic Type을 활용하여 시스템 설정에 따라 앱 글자 크기가 변경될 수 있도록 구현해요.

## 3️⃣ 추가 구현 사항

- 연락처를 UserDefault에 저장해서 앱을 종료해도 저장했던 정보가 날라가지 않도록 했어요.
- 연락처 셀 클릭시 전화번호 클립보드 복사 후 메시지 보내는 기능 추가했어요.
   - 동일 정보 시 이름, 나이순, 연락처 순으로 정렬해요
- 전화번호 추가 시 키보드가 자동으로 올라오도록 했어요. 그리고 연락처 3개 항목이 다 입력되어야 저장 버튼이 활성화돼요.
- 전화번호 13자리 이상은 입력하지 못하도록 막았어요.
   - 이름 입력 에러는 공백 입력 시 나타나요.
   - 나이 입력 에러는 0 혹은 3자리 수 초과 입력 시 나타나요.
   - 전화번호 입력 에러는 8자리 이하 입력 시 나타나요.
- 사용자 검색 시 키보드를 내릴 수 있는 툴 바 악세사리 뷰를 추가했어요.
- 다크모드와 라이트모드, 가로모드와 세로모드에서 유연하게 반응하도록 했어요.

## 🍀 PR Point

- dynamic type
    - 글씨 크기가 최대로 커졌을때 어쩔 수 없이 `이름` 표시 영역이나 `실제 이름결과` 중 어느곳이 희생되야 하는지 애매했어요. 설명 라벨보다는 실제 데이터가 마지막까지 보이도록 지키는것이 올바른 결정이지 않을까 생각하는데, 그렇게하면 어떤 데이터인지가 모호해서 일단 설명 라벨을 우선순위를 높여놨는데 콘은 어떻게 생각하시는 지 궁금해요. 🤔
    - `performsFirstActionWithFullSwipe` 를 `false` 로 정의 해놨는데 데이터를 삭제하는 과정의 허들을 높여서 데이터 삭제가 즉각적이지 못하도록 하였는데 이런점에 대한 이러한 부분에 대해 현업의 표준이 있는지 궁금해요~~!

sloDevz committed:
feat: AddContactViewController와 AddContact 스토리보드 추가, ContactListViewController에 NavigationController 추가 및 add 버튼추가

sloDevz committed:
feat: AddContact 스토리보드 UI 구성

sloDevz committed:
feat: ContactList 와 AddContact 스토리보드 연결

sloDevz committed:
feat: ContactListVC에 싱글턴DataManager 생성

sloDevz committed:
feat: Validator에 유저의 PhoneNumber 입력값을 형식에 맞게 고치는 메서드 추가

sloDevz committed:
feat: Validator를 위한 UnitTest 생성

sloDevz committed:
chore: UnitTest 주석삭제

sloDevz committed:
test: Validator UnitTest추가

sloDevz committed:
Merge pull request #1 from sloDevz/step2

Step2

sloDevz committed:
feat: AddContactViewController Cancel,Save 버튼메서드 및 TextField변수 생성

devxsby committed:
merge: main 브랜치 최신화 및 conflict 해결

sloDevz committed:
fix: ArrayExtension Target 추가

devxsby committed:
feat: ContactUIManager 대략적인 구조 추가

sloDevz committed:
feat: AddContactViewController 저장기능 구현

sloDevz committed:
feat: runProgram()에 @discardableResult키워드 추가 및 Errors수정 컨벤션수정

devxsby committed:
chore: ContactUIManager 안 validator 프로토콜 정리

devxsby committed:
feat: 연락처 추가 뷰에서 취소 버튼 클릭 시 띄우는 alert controller 추가

devxsby committed:
feat: requestValidation 추가

- 기존의 age, phoneNumber만 검출하는 로직에서 name도 검출하도록 변경했어요.

sloDevz committed:
feat: Save시Alert 구현 및 변수명 변경

sloDevz committed:
chore: 불필요한 코드 제거

sloDevz committed:
feat: TextField Delegate 추가 및 연락처TextField 실시간 형식변경 구현

devxsby committed:
fix: phone number 유효 카운트 검출 수 변경

devxsby committed:
chore: 불필요한 함수 호출 제거

sloDevz committed:
feat: Save전 이름 공백제거 기능, phonNumber 13자리 최대값제한 기능(미완성)

devxsby committed:
feat: 전화번호 13자리 이상일 시 입력을 제한하는 메서드 추가

devxsby committed:
chore: AddContact 스토리보드 연결 방식 modal에서 코드로 변경

devxsby committed:
chore: contactUIManager 객체 주입 방식 변경

sloDevz committed:
feat: Save후 전 화면으로 돌아갔을때 TableView 최신화 기능 구현

devxsby committed:
chore: Validator Unit Test 삭제

devxsby committed:
chore: appdelegate, scenedelegate 주석 삭제

devxsby committed:
feat: contact list count 하는 메서드 추가

devxsby committed:
feat: text field가 모두 채워져야 연락처 저장 버튼이 활성화 되도록 구현

devxsby committed:
feat: TableViewCell row를 밀어서 삭제하는 메서드만 추가

삭제하는 로직 처리는 userdefault에 연락처를 저장하는 기능 구현 후 하겠습니다.

sloDevz committed:
feat: UserDefault를 이용한 저장기능 구현
- 테스트를 위한 코드도 포함됐어요 (이름이 Outsider 일때 저장하면 모든데이터 제거돼요)

sloDevz committed:
feat: UserDefault Test코드 수정

sloDevz committed:
feat: 삭제기능 추가

sloDevz committed:
feat: 커스텀셀 구현

sloDevz committed:
feat: customed Cell의 identifier식별의 안전성을 위한 UITableViewCell extension 추가

sloDevz committed:
feat: dequeueReusableCell의 안전성을 위한 UITableView extension 추가 및 그에따른 로직 변경

sloDevz committed:
feat: searchBar 추가

devxsby committed:
chore: Validator scheme 제거

sloDevz committed:
Merge branch 'loganDev' into step3

devxsby committed:
chore: 컨벤션 정리

devxsby committed:
chore: 주석 변경

sloDevz committed:
feat: searchResultViewController 추가 및 UISearchBarDelegate 준수

sloDevz committed:
feat: Person 이름 나이순 정렬 추가

devxsby committed:
feat: 폴더링

devxsby committed:
chore: 폴더 구조 및 파일 위치 변경

sloDevz committed:
feat: searchBar 임시 구현

devxsby committed:
chore: 접근 제한 수정 및 cell cetdata 메서드 추가

devxsby committed:
feat: user search 기능 추가

devxsby committed:
feat: 불필요한 코드 삭제 및 변수명 변경

devxsby committed:
feat: 연락처 셀 클릭 시 해당 전화번호로 메시지 보내는 기능 추가

sloDevz committed:
feat: DynamicText 적용

devxsby committed:
fix: 사용자 검색 시 데이터 소스 오류 해결

devxsby committed:
chore: Contact Table View Cell 레이아웃 조정

devxsby committed:
feat: name textfield becomeFirstResponder 추가

sloDevz committed:
feat: SearchBar Cancel Button 없애고 키보드 ToolBar의 Done버튼 생성 키보드 내리기 구현

devxsby committed:
chore: 불필요한 코드 삭제 및 메서드 명 변경

protocorn93 commented on ContactManager/ContactManager/Data/Model/Person.swift:
```diff
@@ -0,0 +1,26 @@
+        } else {
+            return lhs.phoneNum < rhs.phoneNum
+        }
+    }
```
비교 연산자를 직접 구현해주신 점 인상깊어요! 👍 다만 저는 보통 이렇게 정렬은 위한 코드는 외부에서 호출시 `sorted { }` 안에 위치시키곤해요. 정렬 기준이 가져다 사용하는 곳 마다 다를 수 있으니까요!

protocorn93 commented on ContactManager/ContactManager/Data/Service/DataManager.swift:
```diff
@@ -11,9 +11,9 @@ final class DataManager {
-    private var contacts: Set<Person> = Set<Person>()
-    
-    private init() {}
+    var contacts: Set<Person> = Set<Person>()
```
메서드로 인터페이스를 제공했는데 `private` 접근제어자는 왜 제거되었나요?

protocorn93 commented on ContactManager/ContactManager/Data/Service/DataManager.swift:
```diff
@@ -23,8 +23,22 @@ final class DataManager {
         return contacts
     }
     
+    func getcontactsDataAsPerson() -> [Person] {
```
1. `getContacts-`가 되어야 할 것 같아요.
2. 구현부를 보면 단순히 `getContacts`를 해도 무방할 것 같은데 네이밍을 이렇게 결정한 이유가 궁금해요~ 외부에선 자료구조에 대해 알고 있을 필요는 없어보여요.

protocorn93 commented on ContactManager/ContactManager/Data/Service/DataManager.swift:
```diff
@@ -23,8 +23,22 @@ final class DataManager {
+        return sortedDatas
+    }
+    
     func getContactsList() -> [String] {
```
`DataManager`가 포맷팅까지 해줄 필요가 있을까 싶긴해요! `DataManager`는 단순히 배열을 반환하고 외부에서 사용하는 곳에 맞춰 포맷팅을 해주는 것이 좀 더 적절한 역할 배분인 것 같아요!

protocorn93 commented on ContactManager/ContactManager/Data/Service/InputEditor.swift:
```diff
@@ -47,7 +47,7 @@ struct InputEditor: InputEditorProtocol {
     func requestValidation(with userInputModel: UserInputModel) throws -> Person {
         do {
-            let person = try validator.checkValidAgeAndNum(input: userInputModel)
+            let person = try validator.checkValidFormat(input: userInputModel)
```
포맷을 검증한다는 의미에서 더 적절한 네이밍으로 잘 수정해주신 것 같아요 👍

protocorn93 commented on ContactManager/ContactManager/Data/Service/Validator.swift:
```diff
@@ -31,13 +32,17 @@ struct Validator: ValidatorProtocol {
     private func validateNumber(_ number: String) throws {
         let numberSplit = number.split(separator: "-").map { String($0) }
-        if number.count < 10 || numberSplit.count < 3 { throw Errors.wrongPhoneNumber }
+        if number.count < 11 || numberSplit.count < 3 { throw Errors.wrongPhoneNumber }
```
이런 상수는 Validator 선언부 상위에 위치시켜서 관리하곤해요. 

```swift 
class Validator { 
  
  // MARK: Constants 
  enum Constants { 
    static let maxCountOfNumbers: Int = 11
  }
}
```

이렇게요!

protocorn93 commented on ContactManager/ContactManagerUI/Presentation/ContactList/View/ContactTableViewCell.swift:
```diff
@@ -0,0 +1,21 @@
+    @IBOutlet weak private var age: UILabel!
+    @IBOutlet weak private var phoneNumber: UILabel!
+    
+    func setData(with person: Person) {
```
살짝 욕심을 더해보자면 보통 전 이럴 때 뷰모델을 사용하곤해요. UI는 Person이라는 도메인 모델을 모르게하고, 단순히 화면에 보여질 값들만 담는 뷰모델만 알고 있는거죠! 

```swift 
struct ContactTableViewModel { 
  let name: String 
  let age: String
  let phoneNumber: String
}

func setData(_ viewModel: ContactTableViewModel) {
  name.text = viewModel.name 
  age.text = viewModel.age 
  phoneNumber = viewModel.phoneNumber
}
```

protocorn93 commented on ContactManager/ContactManagerUI/Presentation/ContactList/View/ContactTableViewCell.swift:
```diff
@@ -0,0 +1,21 @@
+    
+    @IBOutlet weak private var name: UILabel!
+    @IBOutlet weak private var age: UILabel!
+    @IBOutlet weak private var phoneNumber: UILabel!
```
UI 컴포넌트에는 해당 컴포넌트를 유추할 수 있는 네이밍을 가져가는게 좋아요! 

```swift 
nameLabel: UILabel! 
ageLabel: UILabel! 
```

protocorn93 commented on ContactManager/ContactManagerUI/Presentation/ContactList/Controller/ContactListViewController.swift:
```diff
@@ -0,0 +1,143 @@
+        let bar = UIToolbar()
+        let done = UIBarButtonItem(title: "닫기", style: .done, target: self, action: #selector (closeButtonTappedAction))
+        bar.items = [done]
+        bar.sizeToFit()
```
`sizeToFit`을 호출해주신 이유가 무엇일까요?

protocorn93 commented on ContactManager/ContactManagerUI/Global/Extension/UserDefaults+.swift:
```diff
@@ -0,0 +1,14 @@
+
+extension UserDefaults {
+    enum Keys {
+        static var contacts = "contactsData"
```
`static let`으로 선언되어도 무방해보여요

protocorn93 commented on ContactManager/ContactManagerUI/Presentation/AddContact/Controller/AddContactViewController.swift:
```diff
@@ -0,0 +1,126 @@
+    // MARK: - Properties
+
+    var contactUIManager: ContactUIManager?
+    var contactListTableView: UITableView?
```
이렇게 프로퍼티에 직접 접근을 통해 의존성을 주입해주는 방법을 프로퍼티 의존성 주입이라고 하는데요. 물론 이런 방법도 종종 사용하지만 스토리보드를 사용하면서 생성자 의존성 주입을 하는 방법에 대해서도 한번 찾아보시면 좋을 것 같아요.

protocorn93 replied:
`reloadData`를 호출하기 위해 `contactListTableView`를 직접 알고 있는데요. `AddContactViewController`는 자신의 앞에 테이블뷰가 있었는지, 콜렉션뷰가 있었는지, 어떤 형태의 UI가 있었는지 알 필요가 없는게 좋아요. 델리게이트 패턴을 활용해볼 수 있을 것 같아요. 

`AddContactViewController`는 단순히 델리게이트에게 "난 연락처 추가를 완료했어" 라는 메시지만 전달할 수 있으면 될 것 같아요

protocorn93 commented on ContactManager/ContactManagerUI/Presentation/AddContact/Controller/AddContactViewController.swift:
```diff
@@ -0,0 +1,126 @@
+    @IBOutlet weak private var nameTextField: UITextField!
+    @IBOutlet weak private var ageTextField: UITextField!
+    @IBOutlet weak private var phoneNumberTextField: UITextField!
+    @IBOutlet weak var saveButton: UIBarButtonItem!
```
이 프로퍼티만 `internal`로 선언된 이유가 무엇일까요?

protocorn93 commented on ContactManager/ContactManagerUI/Presentation/AddContact/Controller/AddContactViewController.swift:
```diff
@@ -0,0 +1,126 @@
+            contactListTableView?.reloadData()
+            self.dismiss(animated: true)
+        } catch {
+            presentErrorAlert(with: error.localizedDescription)
```
알럿을 통해 에러 케이스를 사용자에게 알려주려고 한 점 좋은 것 같아요👍

protocorn93 commented on ContactManager/ContactManagerUI/Presentation/AddContact/Controller/AddContactViewController.swift:
```diff
@@ -0,0 +1,126 @@
+        guard let contactUIManager else { return }
+        
+        do {
+            guard let name = nameTextField.text else { throw Errors.wrongName }
+            guard let age = ageTextField.text else { throw Errors.wrongAge }
+            guard let phoneNumber = phoneNumberTextField.text else { throw Errors.wrongPhoneNumber }
+            
+            let trimmedName = name.split(separator: " ").joined()
+            let userInputModel = UserInputModel(name: trimmedName, age: age, phoneNum: phoneNumber)
+            
+            try contactUIManager.runProgram(menu: .add, userInputModel: userInputModel)
+            contactUIManager.setStoredContactsData()
+            contactListTableView?.reloadData()
+            self.dismiss(animated: true)
```
이 안에 UI 접근, 비즈니스 로직, 데이터 저장, 화면 전환이 모두 위치하고 있는데, 맛있는 코드 과정을 진행하면서 이것들을 분리하는 방법을 학습해가시면 좋을 것 같아요!

protocorn93 commented on ContactManager/ContactManagerUI/Presentation/AddContact/Controller/AddContactViewController.swift:
```diff
@@ -0,0 +1,126 @@
+
+extension AddContactViewController: UITextFieldDelegate {
+    
+    func textFieldDidChangeSelection(_ textField: UITextField) {
```
[textField(_:shouldChangeCharactersIn:replacementString:)](https://developer.apple.com/documentation/uikit/uitextfielddelegate/1619599-textfield) 이 메서드를 활용하면 더 정교한 입력값 조정이 가능해요!

protocorn93 commented on ContactManager/ContactManagerUI/Presentation/AddContact/Controller/AddContactViewController.swift:
```diff
@@ -0,0 +1,126 @@
+        checkSaveButtonState()
+    }
+    
+    func checkMaxLength(_ textField: UITextField, _ maxLength: Int) {
```
`private` 접근 제어자가 필요할 것 같아요

protocorn93 commented on ContactManager/ContactManagerUI/Presentation/ContactList/Controller/ContactListViewController.swift:
```diff
@@ -0,0 +1,143 @@
+    
+    // MARK: - @IBAction Properties
+    
+    @IBAction func tappedAddContactButton(_ sender: UIBarButtonItem) {
```
마찬가지로 `private` 접근제어자가 필요할 것 같아요.

protocorn93 commented on ContactManager/ContactManagerUI/Presentation/ContactList/Controller/ContactListViewController.swift:
```diff
@@ -0,0 +1,143 @@
+    // MARK: - @IBAction Properties
+    
+    @IBAction func tappedAddContactButton(_ sender: UIBarButtonItem) {
+        guard let addContactVC = UIStoryboard(name: "AddContact", bundle: nil).instantiateViewController(withIdentifier:"AddContactViewController") as? AddContactViewController else { return }
```
UIStoryboard의 [instantiateInitialViewController(creator:)](https://developer.apple.com/documentation/uikit/uistoryboard/3213988-instantiateinitialviewcontroller)이 생성자를 통해 의존성을 생성 시점에 주입해줄 수 있어요~ 활용해보시는 것도 좋을 것 같아요!

protocorn93 reviewed (changes requested):
두 분 모두 쉽지 않은 프로젝트였을텐데 고생하셨어요! 보너스 스탭까지 모두 잘 구현해주셨어요! 👍

--- 

## 질문
> dynamic type
글씨 크기가 최대로 커졌을때 어쩔 수 없이 이름 표시 영역이나 실제 이름결과 중 어느곳이 희생되야 하는지 애매했어요. 설명 라벨보다는 실제 데이터가 마지막까지 보이도록 지키는것이 올바른 결정이지 않을까 생각하는데, 그렇게하면 어떤 데이터인지가 모호해서 일단 설명 라벨을 우선순위를 높여놨는데 콘은 어떻게 생각하시는 지 궁금해요. 🤔

이건 정답이 없는 것 같아요. 다이나믹 타입에 대응해서 최대한 유연하게 레이아웃을 구성하는게 최우선일 것 같아요. 레이아웃 요소(텍스트의 길이와 폰트 크기)에 따라 셀의 높이를 동적으로 조정해줄 수 있어야 해요. 이건 Dynamic Cell Height 키워드로 검색해보시면 많은 자료들이 존재해요. 

만일 한정된 레이아웃이라면 분명 어느 하나는 안보이게 될텐데, 그러면 보이스오버를 지원해주는게 하나의 방법이 될 수 있지 않을까 싶네요! 

> performsFirstActionWithFullSwipe 를 false 로 정의 해놨는데 데이터를 삭제하는 과정의 허들을 높여서 데이터 삭제가 즉각적이지 못하도록 하였는데 이런점에 대한 이러한 부분에 대해 현업의 표준이 있는지 궁금해요~~!

이것 역시 정답은 없어요. 어떤 사용자는 스와이프 액션에 대해 인지하고 있을 것이고 어떤 사용자는 아닐 것이기 때문이에요. 의외로 사용자들은 복잡한 제스처는 하지 않아요. 최대한 탭 이벤트로 모든걸 제공할 수 있는 것이 보다 쉬운 UX라고 생각해요. 

앱을 만들고 서비스를 만든다면 대상자가 얼마나 잘 알고 사용하는지에 대해 고민해보는게 좋아요. 정답은 없는 것 같아요!

sloDevz committed:
refactor: 코드리뷰 반영 1차
- Person의 Comparable 준수 해제 및 정렬을 위한 Sorted 로직은 외부에서 직접 구현
- DataManager의 contacts 접근제한자 private로 다시 변경
- contacts 접근제한자 변경으로 인한 메서드추가 -> 최초실행시 저장된 데이터를 불러오기 위한 인터페이스 제공
- DataManager의 getcontactsDataAsPerson메서드명 -> getContacts 로 변경
- DataManager에서 sorted() 사용하여 정렬된 contacts배열을 얻는 로직 위치를 ContactUIManager의 getContactsData의 이름을 getSortedContacts로 바꾸고 이동
- Validator에 사용되는 enum Constants를 Validator 중첩타입으로 Validator최상단으로 옮기고 기존의 Constants 파일은 제거
- Cell 구성하는 과정에 ViewModel 사용 -> Model폴더에 ContactTableViewModel 생성 및 ContactTableViewCell의 setData 로직 약간 변경
- 컴포넌트의 유추를 위해 ContactTableViewCell 의 IBOutlet 변수들 이름변경
- UserDefaults의 contacts 상수로 변경
- AddContactViewController의 saveButton 접근제한자 private
- ContactListViewController의 tappedAddContactButton 접근제한자 private
- checkMaxLength 메서드 접근제한자 private

sloDevz committed:
refactor: delegate 패턴 사용을 위한 UdateDataProtocol 생성

sloDevz committed:
refactor: delegate 패턴으로 ContactListViewController와 AddContactViewController간의 데이터 전달구현
- AddContactViewController의 contactUIManger와 contactListTableVeiw 제거
- AddContactViewController에 delegate 전달자 추가해줬어요
- 기존에 tappedSaveButton에 불필요하게 구성중이던 부분들이 제거됨에 따라서 delegate전달자가 그 일을 대신해요
- contactUIManager가 사라져서 formmatingPhoneNumber메서드를 사용하기 위해 위치를 AddContactViewController로 옮겼어요
- AddContactViewController에서 얼럿창을 띄우기 위해 에러를 받아야 하기 때문에 UpdateDataProtocol의 delevaryUpdateData 메서드가 에러 반환가능한 메서드로 정의해주었어요

sloDevz committed:
refactor: 불필요한 메서드 호출 제거

sloDevz committed:
feat: shouldChangeCharactersIn 추가

sloDevz committed:
feat: 나이와 연락처는 숫자만 입력되도록 구현

sloDevz committed:
refactor: shouldChangeCharactersInt 내부로직 수정

sloDevz committed:
refactor: AddContactViewController initializer 정의

sloDevz committed:
refactor: 정의된 initializer 적용
- creator 생성자를 통해 의존성을 생성 시점에 주입했어요

sloDevz committed:
style: 공백제거 및파일내 코드 위치 변경

sloDevz committed:
refactor: ContactTableViewModel 제거 및 코드변경
- ContactTableViewModel이 UserInputModel로 대체가능하기 때문에 제거했어요

sloDevz committed:
chore: delegate선언을 let으로 변경

protocorn93 commented on ContactManager/ContactManagerUI/Presentation/AddContact/Controller/AddContactViewController.swift:
```diff
@@ -11,7 +11,7 @@ final class AddContactViewController: UIViewController {
     // MARK: - Properties
     
-    var delegate: UpdateDataProtocol
+    let delegate: UpdateDataProtocol
```
델리게이트는 weak var로 참조하는게 좋을 것 같아요. 쌍방 참조가 발생할 확률이 높아요.

devxsby committed:
chore: weak var delegate 선언
